jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c

```diff
@@ -317,180 +318,129 @@ static ecma_value_t
+  ecma_deref_object (new_regexp_obj);
+  ecma_deref_object (func_obj);
 
-    ecma_value_t this_index_number = ecma_op_to_number (this_index_value);
+  return result;
+#else /* !ENABLED (JERRY_ES2015) */
 
-    if (ECMA_IS_VALUE_ERROR (this_index_value))
-    {
-      ecma_free_value (this_index_value);
-      goto cleanup;
-    }
+  if (ecma_object_is_regexp_object (regexp_arg))
+  {
+    return ecma_regexp_match_helper (regexp_arg, this_to_string_value);
+  }
 
-    ecma_number_t this_index = ecma_get_number_from_value (this_index_number);
+  ecma_string_t *pattern_p;
 
-    /* 8.f.iii.2. */
-    if (this_index == previous_last_index)
-    {
-      /* 8.f.iii.2.a. */
-      ecma_value_t index_put_value = ecma_op_object_put (regexp_obj_p,
-                                                         ecma_get_magic_string (LIT_MAGIC_STRING_LASTINDEX_UL),
-                                                         ecma_make_number_value (this_index + 1),
-                                                         true);
-      if (ECMA_IS_VALUE_ERROR (index_put_value))
-      {
-        ecma_free_value (this_index_value);
-        ecma_free_number (this_index_number);
-        goto cleanup;
-      }
+  if (ecma_is_value_undefined (regexp_arg))
+  {
+    pattern_p = ecma_get_magic_string (LIT_MAGIC_STRING__EMPTY);
+  }
+  else
+  {
+    pattern_p = ecma_op_to_string (regexp_arg);
 
-      /* 8.f.iii.2.b. */
-      previous_last_index = this_index + 1;
-    }
-    else
+    if (JERRY_UNLIKELY (pattern_p == NULL))
     {
-      /* 8.f.iii.3. */
-      previous_last_index = this_index;
+      return ECMA_VALUE_ERROR;
     }
-
-    /* 8.f.iii.4. */
-    JERRY_ASSERT (ecma_is_value_object (exec_value));
-    ecma_object_t *exec_obj_p = ecma_get_object_from_value (exec_value);
-
-    ecma_value_t match_string_value = ecma_op_object_get (exec_obj_p, index_zero_string_p);
-    JERRY_ASSERT (!ECMA_IS_VALUE_ERROR (match_string_value));
-
-    /* 8.f.iii.5. */
-    ecma_value_t completion;
-    completion = ecma_builtin_helper_def_prop_by_index (new_array_obj_p,
-                                                        n,
-                                                        match_string_value,
-                                                        ECMA_PROPERTY_CONFIGURABLE_ENUMERABLE_WRITABLE);
-
-    JERRY_ASSERT (ecma_is_value_true (completion));
-    /* 8.f.iii.6. */
-    n++;
-
-    ecma_free_value (match_string_value);
-    ecma_free_value (this_index_value);
-    ecma_free_number (this_index_number);
-
-    ecma_deref_object (exec_obj_p);
   }
 
-  if (n == 0)
+  ecma_value_t new_regexp = ecma_op_create_regexp_object (pattern_p, 0);
+
+  if (ECMA_IS_VALUE_ERROR (new_regexp))
   {
-    /* 8.g. */
-    ret_value = ECMA_VALUE_NULL;
+    ecma_deref_ecma_string (pattern_p);
+    return new_regexp;
   }
-  else
+
+  ecma_value_t result = ecma_regexp_match_helper (new_regexp, this_to_string_value);
+
+  if (ECMA_IS_VALUE_ERROR (result))
```

Comment:
I don't think we need an error check here, we would just return with the result anyway.